org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Scanner.java

```diff
@@ -2243,13 +2126,33 @@ protected int scanForTextBlock() throws InvalidInputException {
 		}
 		if (lastQuotePos > 0)
 			this.currentPosition = lastQuotePos;
-		this.currentPosition = (lastQuotePos > 0) ? lastQuotePos : this.startPosition + this.rawStart;
+		this.currentPosition = (lastQuotePos > 0) ? lastQuotePos : this.startPosition + this.textBlockOffset;
 		throw unterminatedTextBlock();
 	} catch (IndexOutOfBoundsException e) {
-		this.currentPosition = (lastQuotePos > 0) ? lastQuotePos : this.startPosition + this.rawStart;
+		this.currentPosition = (lastQuotePos > 0) ? lastQuotePos : this.startPosition + this.textBlockOffset;
 		throw unterminatedTextBlock();
 	}
 }
+
+private int getAndStoreStringFragment(int lastFragmentStart, int textFragmentEnd) {
+	if (this.jumpingOverEmbeddedExpression == 0) {
+		addStringTemplateComponent(new TextFragment(lastFragmentStart, textFragmentEnd - 1, getCurrentTokenInRange(lastFragmentStart - this.startPosition, textFragmentEnd - this.startPosition)));
+		int eeStart = this.currentPosition;
+		jumpOverEmbeddedExpression();
+		// verify that we are at } or handle error
+		int eeEnd;
```

Comment:
Do we handle the case of missing } properly ?

Reply:
I will add some validation and new error reporting. But doubt it will be of much use as it's not common to not have a closing } after a template expression. The scanner usually ends up with one of the closing } of  the next block/method and subsequently reports an unterminated text block error.